### PR TITLE
add options.headingObjectCallback allows better customization

### PR DESCRIPTION
**Pull request by k2s**
This makes parsing more customizable. Headings doesn't need now to be `H` elements.
It is possible to set level and label based on original node properties.

It also fixes problem with duplicate calls to `headingLabelCallback()`.

It fixes my ticket in  #163.

### CODE REVIEW

File: src/js/parse-content.js [hunk 1]
```diff
@@ -44,6 +53,13 @@ module.exports = function parseContent (options) {
       obj.childNodes = heading.childNodes
     }
 
+    if (options.headingObjectCallback) {
```
Comment (tscanlin) [inline src/js/parse-content.js]: can we change this to: 
```
if (options.headingObjectCallback) {
  return options.headingObjectCallback(obj, heading)
}
```

File: README.md [hunk 2]
```diff
@@ -201,7 +201,12 @@ skipRendering: false,
 // Optional callback to change heading labels. 
 // For example it can be used to cut down and put ellipses on multiline headings you deem too long.
 // Called each time a heading is parsed. Expects a string in return, the modified label to display.
-headingLabelCallback: function (string) => string
+headingLabelCallback: function (string) => string,
+// ignore headings that are hidden in DOM
+ignoreHiddenElements: false,
+// Optional callback to modify properties of parsed headings
+// The heading will be excluded from TOC of this callback returns FALSE
+headingObjectCallback: function (object, HTMLElement) => false | void
```
Comment (tscanlin) [inline README.md]: then if we are returning a new value from the function this would probably change to:
`function (object, HTMLElement) => object | null`

**Comment by k2s** [comment]
Yes, your suggestions are much cleaner.

What I was thinking was that it should not accept changes to all obj properties. That we should pass copy of obj and merge back to obj only whitelisted properties.
Wanted to ask you about this if you will consider to accept my changes.

There are currently properties: id, children, nodeName, headingLevel, and textContent
Not sure if it will be safe if user will modify for example `children` property.

Also didn't came a cross the need to modify the HTMLElement, but should be also evaluated if it would not be helpfull in some cases to let the TOC point to another(parent) element.

I commited your proposal, but please think about my other comments and let me know if I should make the callback more safe by merging limited properties set to obj.

**Comment by tscanlin** [comment]
Thanks for making the change. I appreciate your thoughtfulness in thinking through this feature. I think it’s ok to let all object properties get merged though. It’s a powerful feature but the extra code and complexity needed for it to be safer probably won’t be necessary for most use cases.

**Comment by tscanlin** [comment]
`4.10.0` is published with this change. Thanks again for the PR!

**Pull request merged**